### ISSUE TITLE
Ban empty for loop iterator

### ISSUE THREAD

**Issue by delanym** [checkstyle/checkstyle#14934]
As of today Eclipse and Intellij IDEs cannot be made to format code in the same way.

The for loop with an empty iterable is an outstanding problem
https://youtrack.jetbrains.com/issue/IDEA-288050

To get around this tragedy, Intellij users must rely on an Eclipse formatter bundled in the plugin
https://plugins.jetbrains.com/plugin/6546-adapter-for-eclipse-code-formatter

But this plugin currently cannot format all language constructs in Java 21
https://github.com/krasa/EclipseCodeFormatter/issues/269

Checkstyle can enforce how empty iterators should be formatted, but there is no way to outright ban them.
https://checkstyle.sourceforge.io/checks/whitespace/emptyforiteratorpad.html

A for loop with an empty iterator is not a pretty thing and there are better alternatives. For example,
```java
for (;;)
```
can be
```java
while(true)
```
and
``` java
for (i = props.getChildProperties().keySet().iterator(); i.hasNext();) {
```
can become either
```java
var i = props.getChildProperties().keySet().iterator();
while (i.hasNext()) {
```
or 
```java
public static Iterable<Properties> iterate(Properties node) return () -> (Iterator<Properties>) node.children().asIterator();

...

for (var child : iterate(props)) {
```

When the proposed rule is applied only one form of the classic for loop will be allowed: it must have an initializer, a condition, and an iterator.

If IDEs cannot accommodate each other, then the form the code takes will need to be restricted.

**Comment by romani** [comment]
>for (;;) can be while(true)

This  can be covered by Checkstyle. But instead of writing whole new rule for it, you can make your own rule by xpath matching https://checkstyle.org/checks/coding/matchxpath.html#MatchXpath
When you end up with config that works, please share it here to be reused by others 


>for (i = props.getChildProperties().keySet().iterator(); i.hasNext();) {

I don't think checkstyle is capable to handle it reliably. There are types that  matter in this context and we are not aware about types https://checkstyle.org/writingchecks.html#Limitations
